subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/AbstractTransformer.java

```diff
@@ -18,27 +18,96 @@
+import org.gradle.internal.isolation.Isolatable;
 
+import javax.annotation.Nullable;
 import java.io.File;
 
-public abstract class AbstractTransformer<T> implements Transformer {
+public abstract class AbstractTransformer<T, P> implements Transformer {
     private final Class<? extends T> implementationClass;
-    private final HashCode inputsHash;
     private final ImmutableAttributes fromAttributes;
+    private final DomainObjectContextProjectStateHandler projectStateHandler;
+    private IsolatableParameters<P> isolatableParameters;
+    private final ProjectStateRegistry.SafeExclusiveLock isolationLock;
+    private WorkNodeAction isolateAction;
 
-    public AbstractTransformer(Class<? extends T> implementationClass, HashCode inputsHash, ImmutableAttributes fromAttributes) {
+    public AbstractTransformer(Class<? extends T> implementationClass, ImmutableAttributes fromAttributes, DomainObjectContextProjectStateHandler projectStateHandler) {
         this.implementationClass = implementationClass;
-        this.inputsHash = inputsHash;
         this.fromAttributes = fromAttributes;
+        this.projectStateHandler = projectStateHandler;
+        this.isolationLock = projectStateHandler.newExclusiveOperationLock();
+        this.isolateAction = new WorkNodeAction() {
+            @Nullable
+            @Override
+            public Project getProject() {
+                return projectStateHandler.maybeGetOwningProject();
+            }
+
+            @Override
+            public void run() {
+                isolateExclusively();
+            }
+        };
+    }
+
+    static class IsolatableParameters<T> {
+        private HashCode secondaryInputsHash;
+        private Isolatable<T> isolatableParameters;
+
+        public IsolatableParameters(Isolatable<T> isolatableParameters, HashCode secondaryInputsHash) {
+            this.secondaryInputsHash = secondaryInputsHash;
+            this.isolatableParameters = isolatableParameters;
+        }
+
+        public HashCode getSecondaryInputsHash() {
+            return secondaryInputsHash;
+        }
+
+        public Isolatable<T> getIsolatableParameters() {
+            return isolatableParameters;
+        }
     }
 
     @Override
     public ImmutableAttributes getFromAttributes() {
         return fromAttributes;
     }
 
+    abstract protected IsolatableParameters<P> doIsolateParameters();
+
+    @Override
+    public void visitDependencies(TaskDependencyResolveContext context) {
+        context.add(isolateAction);
+    }
+
+    @Override
+    public void isolateParameters() {
+        if (isolatableParameters == null) {
+            if (!projectStateHandler.hasMutableProjectState()) {
+                projectStateHandler.withLenientState(this::isolateExclusively);
+            } else {
+                isolateExclusively();
+            }
+        }
+    }
+
+    private void isolateExclusively() {
+        isolationLock.withLock(() -> {
```

Comment:
I don't think we necessarily need to do any locking here, this shouldn't ever run concurrently with anything else

Reply:
As discussed, isolation of the transform is not called when we resolve the dependency graph (which already has the project lock), but when we try to run it. And that requires additional locking since different threads can try to run the same registered transform.